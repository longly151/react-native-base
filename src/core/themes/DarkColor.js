module.exports = {
  dark: true,
  colors: {
    primary: '#219EE5',
    background: '#000000',
    card: '#1c1c1e',
    text: '#ffffff',
    border: '#272729',
    notification: '#ff453a',
    grey5: '#393e42',
    grey4: '#43484d',
    grey3: '#5e6977',
    grey2: '#86939e',
    grey1: '#bdc6cf',
    grey0: '#e1e8ee',
    white: '#080808',
    black: '#f2f2f2',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
