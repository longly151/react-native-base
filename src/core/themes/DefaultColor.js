module.exports = {
  dark: false,
  colors: {
    primary: '#219EE5',
    background: '#f2f2f7',
    card: '#ffffff',
    text: '#000000',
    border: '#d8d8d8',
    notification: '#ff3b30',
    grey0: '#393e42',
    grey1: '#43484d',
    grey2: '#5e6977',
    grey3: '#86939e',
    grey4: '#bdc6cf',
    grey5: '#e1e8ee',
    white: '#ffffff',
    black: '#242424',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
