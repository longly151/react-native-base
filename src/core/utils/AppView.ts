import { Font } from '@fonts';
import DefaultColor from '@themes/DefaultColor';
import { Dimensions, StyleProp, ViewStyle, TextStyle, ColorValue, ScaledSize } from 'react-native';
import ViewConfig from '@themes/ViewConfig';

const { width, height } = Dimensions.get('window');

/**
 * Shadow
 */
const androidDepth = {
  umbra: [
    '0px 2px 1px -1px',
    '0px 3px 1px -2px',
    '0px 3px 3px -2px',
    '0px 2px 4px -1px',
    '0px 3px 5px -1px',
    '0px 3px 5px -1px',
    '0px 4px 5px -2px',
    '0px 5px 5px -3px',
    '0px 5px 6px -3px',
    '0px 6px 6px -3px',
    '0px 6px 7px -4px',
    '0px 7px 8px -4px',
    '0px 7px 8px -4px',
    '0px 7px 9px -4px',
    '0px 8px 9px -5px',
    '0px 8px 10px -5px',
    '0px 8px 11px -5px',
    '0px 9px 11px -5px',
    '0px 9px 12px -6px',
    '0px 10px 13px -6px',
    '0px 10px 13px -6px',
    '0px 10px 14px -6px',
    '0px 11px 14px -7px',
    '0px 11px 15px -7px',
  ],
  penumbra: [
    '0px 1px 1px 0px',
    '0px 2px 2px 0px',
    '0px 3px 4px 0px',
    '0px 4px 5px 0px',
    '0px 5px 8px 0px',
    '0px 6px 10px 0px',
    '0px 7px 10px 1px',
    '0px 8px 10px 1px',
    '0px 9px 12px 1px',
    '0px 10px 14px 1px',
    '0px 11px 15px 1px',
    '0px 12px 17px 2px',
    '0px 13px 19px 2px',
    '0px 14px 21px 2px',
    '0px 15px 22px 2px',
    '0px 16px 24px 2px',
    '0px 17px 26px 2px',
    '0px 18px 28px 2px',
    '0px 19px 29px 2px',
    '0px 20px 31px 3px',
    '0px 21px 33px 3px',
    '0px 22px 35px 3px',
    '0px 23px 36px 3px',
    '0px 24px 38px 3px',
  ],
  ambient: [
    '0px 0px 0px 0px',
    '0px 1px 3px 0px',
    '0px 1px 5px 0px',
    '0px 1px 8px 0px',
    '0px 1px 10px 0px',
    '0px 1px 14px 0px',
    '0px 1px 18px 0px',
    '0px 2px 16px 1px',
    '0px 3px 14px 2px',
    '0px 3px 16px 2px',
    '0px 4px 18px 3px',
    '0px 4px 20px 3px',
    '0px 5px 22px 4px',
    '0px 5px 24px 4px',
    '0px 5px 26px 4px',
    '0px 6px 28px 5px',
    '0px 6px 30px 5px',
    '0px 6px 32px 5px',
    '0px 7px 34px 6px',
    '0px 7px 36px 6px',
    '0px 8px 38px 7px',
    '0px 8px 40px 7px',
    '0px 8px 42px 7px',
    '0px 9px 44px 8px',
    '0px 9px 46px 8px',
  ],
};
function interpolate(i: number, a: number, b: number, a2: number, b2: number) {
  return ((i - a) * (b2 - a2)) / (b - a) + a2;
}
function parseShadow(raw: any) {
  const values = raw.split(' ').map((val: any) => +val.replace('px', ''));
  return {
    x: values[0],
    y: values[1],
    blur: values[2],
    spread: values[3], // unused
  };
}
function onDepthChange(depth: number) {
  const s = parseShadow(androidDepth.penumbra[depth]);
  const y = s.y === 1 ? 1 : Math.floor(s.y * 0.5);
  return {
    elevation: depth + 1, // Android
    // iOS
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: y,
    },
    shadowOpacity: Number.parseFloat(interpolate(depth, 1, 24, 0.2, 0.6).toFixed(2)),
    shadowRadius: Number.parseFloat(interpolate(s.blur, 1, 38, 1, 16).toFixed(2)),
  };
}

class CAppView {
  private static _instance: CAppView;

  private constructor() {
    // ...
  }

  public static get Instance(): CAppView {
    /* istanbul ignore else */
    if (!this._instance) {
      this._instance = new this();
    }
    return CAppView._instance;
  }

  fontFamily: Font = 'GoogleSans-Regular';

  safeAreaInsets = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  };

  shadow = (depth: number = 10) => (depth >= 1 ? onDepthChange(depth - 1) : {});

  bottomNavigationBarHeight = 45;

  headerHeight = 50;

  bodyHeight =
    height -
    this.headerHeight -
    this.bottomNavigationBarHeight -
    this.safeAreaInsets.top -
    this.safeAreaInsets.bottom;

  bodyPaddingHorizontal = Number.parseFloat(ViewConfig.spacing.DEFAULT);

  bodyWidth = width - 2 * this.bodyPaddingHorizontal;

  roundedBorderRadius = Number.parseFloat(ViewConfig.borderRadius.DEFAULT);

  bottomSheetBorderRadius = Number.parseFloat(ViewConfig.borderRadius.card);

  itemMarginHorizontal = 10;

  screenWidth = width;

  screenHeight = height;

  isHorizontal = this.screenWidth > this.screenHeight;

  // Carousel
  sliderWidth = this.screenWidth;

  itemWidth = (this.isHorizontal ? 0.35 : 0.75) * this.screenWidth + this.itemMarginHorizontal * 2;

  itemHeight = 300;

  carouselBorderRadius = 8;

  onDimensionChange = (dimensions: { window: ScaledSize; screen: ScaledSize }) => {
    const { window } = dimensions;
    this.screenWidth = window.width;
    this.screenHeight = window.height;
    const isHorizontal = window.width > window.height;
    this.isHorizontal = isHorizontal;

    if (isHorizontal) {
      // Carousel
      this.itemWidth = 0.35 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    } else {
      this.itemWidth = 0.75 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    }
  };

  initSafeArea = (safeAreaInsets: { top: number; bottom: number; left: number; right: number }) => {
    const dimensions = Dimensions.get('window');
    this.safeAreaInsets = safeAreaInsets;
    this.bodyHeight =
      dimensions.height -
      this.headerHeight -
      this.bottomNavigationBarHeight -
      safeAreaInsets.top -
      safeAreaInsets.bottom;
  };

  getHeaderStyle = (theme: typeof DefaultColor) =>
    ({
      headerStyle: {
        backgroundColor: theme.colors.background,
      } as StyleProp<ViewStyle>,
      headerTitleStyle: {
        fontSize: 22,
        color: theme.colors.grey0,
        fontWeight: 'bold',
        fontFamily: AppView.fontFamily,
      } as StyleProp<TextStyle>,
      headerShadowVisible: false,
      headerBackTitleVisible: false,
      headerBackButtonMenuEnabled: false,
      headerTintColor: theme.colors.text,
    } as any);

  getQuickStyle = ({
    outline,
    clear,
    backgroundColor,
    textColor,
    borderColor,
  }: {
    outline?: boolean;
    clear?: boolean;
    backgroundColor: ColorValue;
    textColor: ColorValue;
    borderColor?: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    let quickTextStyles: StyleProp<TextStyle> = {};
    if (outline) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderColor: borderColor || backgroundColor,
        borderWidth: 1,
        elevation: 0,
      };
      quickTextStyles = {
        color: textColor || backgroundColor,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        elevation: 0,
        borderRadius: 0,
      };
      quickTextStyles = { color: backgroundColor };
    } else {
      quickStyles = {
        backgroundColor,
      };
    }
    return { quickStyles, quickTextStyles };
  };

  getInputQuickStyle = ({
    outline,
    clear,
  }: {
    outline?: boolean;
    clear?: boolean;
    textColor: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    if (outline) {
      quickStyles = {
        borderWidth: 1,
        elevation: 0,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderRadius: 0,
        borderBottomWidth: 1,
        elevation: 0,
      };
    } else {
      quickStyles = {
        borderWidth: 1,
      };
    }
    return { quickStyles };
  };
}

const AppView = CAppView.Instance;
export default AppView;
