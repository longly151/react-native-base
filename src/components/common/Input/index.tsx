import React, { useEffect, useImperativeHandle, useMemo, useRef, useState } from 'react';
import {
  ColorValue,
  Keyboard,
  NativeSyntheticEvent,
  Platform,
  StyleProp,
  StyleSheet,
  TextInput as RNTextInput,
  TextInputFocusEventData,
  TextInputProps as RNTextInputProps,
  TextInputSubmitEditingEventData,
  TextStyle,
  View,
  ViewStyle,
} from 'react-native';
import withTailwind from '@components/hoc/withTailwind';
import Icon, { IconProps } from '../Icon';
import AppView from '@utils/AppView';
import { useTheme } from '@react-navigation/native';
import Text, { TextProps } from '../Text';
import _ from 'lodash';
import {
  Control,
  RegisterOptions,
  useController,
  useForm,
  UseFormSetFocus,
  UseFormTrigger,
} from 'react-hook-form';
import CONSTANT from '@configs/constant';

export interface NextFieldProps {
  name: string;
  trigger: UseFormTrigger<any>;
  setFocus: UseFormSetFocus<any>;
}

export interface ReactHookFormProps {
  name?: string;
  control?: Control<any>;
  nextField?: NextFieldProps;
  rules?: RegisterOptions;
}

export interface AdditionalTextInputProps extends ReactHookFormProps {
  title?: string;
  titleProps?: TextProps;
  errorProps?: TextProps;
  leftIconProps?: IconProps;
  rightIconProps?: IconProps;
  color?: ColorValue;
  backgroundColor?: ColorValue;
  inactiveBorderColor?: ColorValue;
  outline?: boolean;
  clear?: boolean;
  component?: any;
}

export interface InputProps extends RNTextInputProps, AdditionalTextInputProps {
  className?: string;
}

const containerStyleKeys = [
  'flex',
  'flexGrow',
  'flexShrink',
  'flexDirection',
  'alignSelf',
  'justifyContent',
  'alignItems',
  'margin',
  'marginLeft',
  'marginRight',
  'marginTop',
  'marginBottom',
  'marginVertical',
  'marginHorizontal',
  'position',
];

const inputContainerStyleKeys = [
  'borderRadius',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderBottomLeftRadius',
  'borderBottomRightRadius',
  'borderWidth',
  'borderTopWidth',
  'borderBottomWidth',
  'borderLeftWidth',
  'borderRightWidth',
  'borderColor',
  'borderTopColor',
  'borderBottomColor',
  'borderLeftColor',
  'borderRightColor',
  'backgroundColor',
  'elevation',
];

export const handleInputStyle = (props: any, theme: App.Theme) => {
  const textColor = props.color || theme.colors.text;
  const backgroundColor =
    props.backgroundColor || (props.outline || props.clear ? 'transparent' : theme.colors.grey5);

  const { quickStyles } = AppView.getInputQuickStyle({
    outline: props.outline,
    clear: props.clear,
    textColor,
  });

  const style: TextStyle = StyleSheet.flatten([
    {
      color: textColor,
      paddingVertical: CONSTANT.INPUT.PADDING_VERTICAL,
      paddingHorizontal: CONSTANT.INPUT.PADDING_HORIZONTAL,
      fontSize: CONSTANT.INPUT.FONT_SIZE,
      fontWeight: CONSTANT.INPUT.FONT_WEIGHT,
      borderRadius: AppView.roundedBorderRadius,
      backgroundColor,
    } as TextStyle,
    quickStyles,
    props.style,
  ]);

  const containerStyle: ViewStyle = _.pick(style, containerStyleKeys) as any;

  const inputContainerStyle: ViewStyle = {
    ...(_.pick(style, inputContainerStyleKeys) as any),
    flexDirection: 'row',
    alignItems: 'center',
  };

  const otherStyle: TextStyle = {
    ...(_.omit(style as any, [...containerStyleKeys, ...inputContainerStyleKeys]) as any),
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    overflow: 'hidden',
    marginTop:
      Platform.OS === 'ios' && props.multiline ? ((style?.paddingVertical || 4) as number) - 4 : 0,
  };

  const titleStyles: TextStyle[] = [
    {
      marginBottom: CONSTANT.INPUT.TITLE_MARGIN_BOTTOM,
      fontSize: CONSTANT.INPUT.TITLE_FONT_SIZE,
    },
    props.titleProps?.style,
  ];

  const errorStyles: TextStyle[] = [
    {
      marginTop: CONSTANT.INPUT.ERROR_MARGIN_TOP,
      fontSize: CONSTANT.INPUT.ERROR_FONT_SIZE,
    },
    props.errorProps?.style,
  ];

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { color: iconColor, fontSize: iconSize, ...otherTextStyle } = otherStyle as any;

  const iconPaddingHorizontal = [
    otherTextStyle.paddingLeft,
    otherTextStyle.paddingRight,
    otherTextStyle.paddingHorizontal,
  ].find(e => e !== undefined);

  const commonIconStyles: StyleProp<TextStyle> = [
    otherTextStyle,
    {
      borderRadius: AppView.roundedBorderRadius,
      padding: otherTextStyle.padding,
      paddingLeft: undefined,
      paddingRight: undefined,
      paddingTop: undefined,
      paddingBottom: undefined,
      paddingHorizontal: iconPaddingHorizontal,
      paddingVertical: undefined,
      flexGrow: undefined,
      flexShrink: undefined,
    },
    props.iconProps?.style,
  ];

  const leftIconStyles: StyleProp<TextStyle> = [
    ...commonIconStyles,
    {
      color: props.leftIconProps?.color || theme.colors.grey3,
      marginRight: -2 * iconPaddingHorizontal + CONSTANT.INPUT.ICON_TITLE_MARGIN_HORIZONTAL,
    },
  ];

  const rightIconStyles: StyleProp<TextStyle> = [
    ...commonIconStyles,
    {
      color: props.rightIconProps?.color || theme.colors.grey3,
      marginLeft: -2 * iconPaddingHorizontal + CONSTANT.INPUT.ICON_TITLE_MARGIN_HORIZONTAL,
    },
  ];

  const inactiveBorderColor = props.inactiveBorderColor || theme.colors.grey5;

  const activeBorderColor =
    inputContainerStyle.borderColor ||
    inputContainerStyle.borderTopColor ||
    inputContainerStyle.borderBottomColor ||
    inputContainerStyle.borderLeftColor ||
    inputContainerStyle.borderRightColor ||
    theme.colors.primary;

  const placeholderTextColor = props.placeholderTextColor || theme.colors.grey3;

  return {
    containerStyle,
    inputContainerStyle,
    titleStyles,
    errorStyles,
    leftIconStyles,
    rightIconStyles,
    otherStyle,
    inactiveBorderColor,
    activeBorderColor,
    placeholderTextColor,
  };
};

const Input: React.FC<InputProps> = React.forwardRef((props, ref: any) => {
  let inputRef = useRef<RNTextInput>(null);

  const theme = useTheme();
  const {
    title,
    titleProps,
    errorProps,
    leftIconProps,
    rightIconProps,
    name,
    onFocus: onFocusProp,
    onBlur: onBlurProp,
    onChangeText: onChangeTextProp,
    onSubmitEditing: onSubmitEditingProp,
    value: valueProp,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    backgroundColor,
    component,
    control,
    nextField,
    rules,
    ...otherProps
  } = props;

  const handledStyle = useMemo(() => handleInputStyle(props, theme), [props, theme]);
  const { placeholderTextColor, inactiveBorderColor, activeBorderColor } = handledStyle;

  const [borderColor, setBorderColor] = useState(
    !props.autoFocus ? inactiveBorderColor : activeBorderColor,
  );
  useEffect(() => {
    setBorderColor(inactiveBorderColor);
  }, [inactiveBorderColor]);

  const { control: formControl } = useForm({ mode: 'onTouched' });
  const { field, fieldState, formState } = useController({
    control: control || formControl,
    defaultValue: otherProps.defaultValue,
    rules,
    name: name || '',
  });

  useEffect(() => {
    const showSubscription = Keyboard.addListener('keyboardDidShow', () => {
      if (Platform.OS === 'android') {
        global.fn?.setIsTabBarHidden?.(true);
      }
    });
    const hideSubscription = Keyboard.addListener('keyboardDidHide', () => {
      if (Platform.OS === 'android') {
        global.fn?.setIsTabBarHidden?.(false);
      }
    });

    return () => {
      showSubscription.remove();
      hideSubscription.remove();
    };
  }, []);

  const onFocus = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    if (!fieldState.error) {
      setBorderColor(activeBorderColor);
    }

    if (Platform.OS === 'android') {
      global.fn?.setIsTabBarHidden?.(true);
    }
    onFocusProp?.(e);
  };

  const onBlur = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    if (!fieldState.error) {
      setBorderColor(inactiveBorderColor);
    }
    if (name && control) {
      field.onBlur();
    }

    if (Platform.OS === 'android') {
      global.fn?.setIsTabBarHidden?.(false);
    }
    onBlurProp?.(e);
  };

  const onChangeText = (e: string) => {
    if (name && control) {
      field.onChange(e);
    }
    onChangeTextProp?.(e);
  };

  const onSubmitEditing = async (e: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => {
    if (!otherProps.multiline) {
      if (name && control && nextField) {
        const isValidated = await nextField?.trigger(name);
        if (isValidated && !formState.touchedFields[nextField.name]) {
          nextField?.setFocus(nextField.name);
        }
      }
    }
    onSubmitEditingProp?.(e);
  };

  useImperativeHandle(ref, () => ({
    ...inputRef.current,
    focus: () => {
      inputRef.current?.focus();
    },
    blur: () => {
      inputRef.current?.focus();
    },
    setNativeProps: (nativeProps: object) => {
      inputRef.current?.setNativeProps(nativeProps);
    },
  }));

  useEffect(() => {
    if (name && control && fieldState.error && borderColor !== theme.colors.error) {
      setBorderColor(theme.colors.error);
    } else if (
      !fieldState.error &&
      inputRef.current?.isFocused() &&
      borderColor !== activeBorderColor
    ) {
      setBorderColor(activeBorderColor);
    }
  }, [borderColor, control, fieldState.error, name, theme.colors.error, activeBorderColor]);

  const renderErrors = () => {
    if (name && control && fieldState.error) {
      return (
        <Text
          testID="ErrorText"
          color={theme.colors.error}
          {...errorProps}
          style={handledStyle.errorStyles}
        >
          {fieldState.error.message}
        </Text>
      );
    }
    return null;
  };

  const handleRef = (e: any) => {
    field.ref(e);
    inputRef = {
      current: e,
    };
  };

  const InputComponent = component || RNTextInput;

  return (
    <View testID="ContainerView" style={handledStyle.containerStyle}>
      {title ? (
        <Text testID="TitleText" {...titleProps} style={handledStyle.titleStyles}>
          {title}
        </Text>
      ) : null}

      <View
        testID="InputView"
        style={[
          handledStyle.inputContainerStyle,
          {
            borderColor,
            borderTopColor: borderColor,
            borderBottomColor: borderColor,
            borderLeftColor: borderColor,
            borderRightColor: borderColor,
          },
        ]}
      >
        {leftIconProps ? (
          <Icon
            testID="LeftIcon"
            color={handledStyle.otherStyle.color}
            size={handledStyle.otherStyle.fontSize}
            {...leftIconProps}
            style={[handledStyle.leftIconStyles, leftIconProps.style]}
          />
        ) : null}
        <InputComponent
          testID="Input"
          ref={handleRef}
          {...otherProps}
          onFocus={onFocus}
          onBlur={onBlur}
          onChangeText={onChangeText}
          onSubmitEditing={onSubmitEditing}
          defaultValue={valueProp || (name && typeof field.value === 'string' ? field.value : '')}
          value={valueProp}
          style={handledStyle.otherStyle}
          placeholderTextColor={placeholderTextColor}
        >
          {props.children}
        </InputComponent>
        {rightIconProps ? (
          <Icon
            testID="RightIcon"
            color={handledStyle.otherStyle.color}
            size={handledStyle.otherStyle.fontSize}
            {...rightIconProps}
            style={[handledStyle.rightIconStyles, rightIconProps.style]}
          />
        ) : null}
      </View>
      {renderErrors()}
    </View>
  );
});

Input.defaultProps = {
  underlineColorAndroid: 'transparent',
};

export interface InputType extends RNTextInput {}

export default withTailwind(Input) as React.FC<InputProps & { ref?: React.LegacyRef<RNTextInput> }>;
