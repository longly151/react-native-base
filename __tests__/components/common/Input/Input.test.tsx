import React from 'react';
import { fireEvent, render, waitFor } from '@utils/TestHelper';
import Input, { InputType } from '@components/common/Input/index';
import { Button, Platform } from 'react-native';
import DefaultColor from '@themes/DefaultColor';
import AppView from '@utils/AppView';
import Color from 'color';
import DarkColor from '@themes/DarkColor';
import { act } from 'react-test-renderer';
import * as Yup from 'yup';
import { Formik } from 'formik';

global.console = {
  ...console,
  error: jest.fn(),
};

let mockTheme = DefaultColor;
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useTheme: jest.fn().mockImplementation(() => mockTheme),
}));

const title = 'Select gender';
const text = 'abc@gmail.com';

beforeEach(() => {
  mockTheme = DefaultColor;
  Platform.OS = 'ios';
});

test('should render correctly', () => {
  const { toJSON, getByTestId } = render(<Input />);
  expect(toJSON()).toMatchSnapshot();
  expect(getByTestId('Input')).toBeTruthy();
});

test('should have working props when unfocusing', () => {
  const { getByTestId, queryByTestId, rerender } = render(
    <Input
      placeholderTextColor={mockTheme.colors.grey4}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500 text-white font-Roboto font-bold italic underline"
    />,
  );

  // title, leftIcon, rightIcon should be null when the relevant props don't exist
  expect(queryByTestId('TitleText')).toBeFalsy();
  expect(queryByTestId('LeftIcon')).toBeFalsy();
  expect(queryByTestId('RightIcon')).toBeFalsy();

  // title, leftIcon, rightIcon should not be null when the relevant props exist
  const RootComponent = (
    <Input
      title="Email"
      leftIconProps={{ name: 'search', size: 24 }}
      rightIconProps={{ name: 'close', size: 24 }}
      placeholderTextColor={mockTheme.colors.grey4}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500 text-white font-Roboto font-bold italic underline"
    />
  );
  rerender(RootComponent);

  expect(getByTestId('TitleText')).toBeTruthy();
  expect(getByTestId('LeftIcon')).toBeTruthy();
  expect(getByTestId('RightIcon')).toBeTruthy();

  // style testing
  expect(getByTestId('ContainerView')).toHaveStyle({
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
  });

  expect(getByTestId('InputView')).toHaveStyle({
    flexDirection: 'row',
    alignItems: 'center',
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderRadius: AppView.roundedBorderRadius,
    borderColor: Color(mockTheme.colors.grey5).darken(0.1).hex(), // unfocus borderColor
  });

  expect(getByTestId('Input')).toHaveProp('placeholderTextColor', mockTheme.colors.grey4);
  expect(getByTestId('Input')).toHaveStyle({
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
    color: 'rgba(255, 255, 255, 1)',
    fontFamily: 'Roboto',
    fontWeight: '700',
    fontStyle: 'italic',
    textDecorationLine: 'underline',
  });

  /**
   * Outline
   */
  rerender(<Input outline />);

  expect(getByTestId('InputView')).toHaveStyle({
    backgroundColor: 'transparent',
  });

  /**
   * Clear
   */
  rerender(<Input clear />);

  expect(getByTestId('InputView')).toHaveStyle({
    backgroundColor: 'transparent',
  });

  /**
   * DarkColor
   */
  mockTheme = DarkColor;
  const { getByTestId: getByTestIdDarkTheme } = render(RootComponent);

  expect(getByTestIdDarkTheme('InputView')).toHaveStyle({
    borderColor: mockTheme.colors.grey5, // unfocus borderColor
  });
});

test('should have working props when focusing', async () => {
  const inputRef: React.RefObject<InputType> = React.createRef();
  const focusIfEmpty = jest.fn();

  const RootComponent = (
    <Input
      ref={inputRef}
      title="Email"
      leftIconProps={{ name: 'search', size: 24 }}
      rightIconProps={{ name: 'close', size: 24 }}
      placeholderTextColor={mockTheme.colors.grey4}
      nextFieldRef={{ current: { focusIfEmpty } }}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500 text-white font-Roboto font-bold italic underline"
    />
  );
  const { getByTestId, rerender } = render(RootComponent);

  /**
   * Input text (Focus Trigger)
   */
  fireEvent(getByTestId('Input'), 'focus');
  expect(getByTestId('InputView')).toHaveStyle({
    borderColor: mockTheme.colors.primary, // focus borderColor
  });
  expect(getByTestId('Input')).toHaveTextContent('');

  // focusIfEmpty nextFieldRef
  fireEvent(getByTestId('Input'), 'onSubmitEditing');
  await waitFor(() => expect(focusIfEmpty).toBeCalledTimes(1));

  // Auto focus
  rerender(<Input autoFocus />);
  expect(getByTestId('InputView')).toHaveStyle({
    borderColor: mockTheme.colors.primary, // focus borderColor
  });
});

test('should work correctly on iOS & Android', () => {
  /**
   * Android
   */
  Platform.OS = 'android';
  const inputRef: React.RefObject<InputType> = React.createRef();

  const RootComponent = <Input ref={inputRef} />;
  const { getByTestId } = render(RootComponent);

  const setIsTabBarHidden = jest.fn();
  global.fn = {
    setIsTabBarHidden,
  };

  fireEvent(getByTestId('Input'), 'focus');
  expect(setIsTabBarHidden).toHaveBeenCalledWith(true);

  fireEvent(getByTestId('Input'), 'blur');
  expect(setIsTabBarHidden).toHaveBeenCalledWith(false);
});

test('should work correctly with formik & nextField', async () => {
  /**
   * Setup Test
   */

  const focusIfEmpty = jest.fn();
  const onChangeText = jest.fn();

  const error = 'Email is required';
  const formValidationSchema = Yup.object().shape({
    email: Yup.string().required(error),
  });

  const onSubmit = jest.fn();

  const inputRef: React.RefObject<InputType> = React.createRef();
  const RootComponent = (
    <Formik
      validationSchema={formValidationSchema}
      initialValues={{
        email: '',
      }}
      onSubmit={onSubmit}
    >
      {formikProps => (
        <>
          <Input
            ref={inputRef}
            formikProps={formikProps}
            name="email"
            title={title}
            onChangeText={onChangeText}
            nextFieldRef={{ current: { focusIfEmpty } }}
          />
          <Button testID="SubmitButton" title="Submit" onPress={formikProps.handleSubmit} />
        </>
      )}
    </Formik>
  );
  const { getByTestId, queryByTestId, findByTestId } = render(RootComponent);

  /**
   * Trigger focusIfEmpty
   */
  fireEvent(getByTestId('Input'), 'focus');
  act(() => inputRef.current?.focusIfEmpty());
  expect(getByTestId('InputView')).toHaveStyle({
    borderColor: mockTheme.colors.primary, // focus borderColor
  });

  /**
   * Trigger Error
   */
  fireEvent.press(getByTestId('SubmitButton'));

  expect(focusIfEmpty).toBeCalledTimes(0);
  expect(onSubmit).toBeCalledTimes(0);
  const errorText = await findByTestId('ErrorText');
  expect(errorText).toHaveTextContent(error);

  /**
   * Successfully Input
   */
  fireEvent.changeText(getByTestId('Input'), text);
  expect(onChangeText).toHaveBeenCalledWith(text);
  expect(inputRef.current?.getText()).toEqual(text);

  fireEvent(getByTestId('Input'), 'onSubmitEditing');

  await waitFor(() => expect(focusIfEmpty).toBeCalledTimes(1));
  expect(queryByTestId('ErrorText')).toBeFalsy();

  fireEvent.press(getByTestId('SubmitButton'));
  await waitFor(() => expect(onSubmit).toBeCalledTimes(1));
  expect(onSubmit).toHaveBeenCalledWith(
    {
      email: text,
    },
    expect.anything(),
  );
});
