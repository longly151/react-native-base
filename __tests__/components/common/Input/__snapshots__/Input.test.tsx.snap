// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<View
  style={Object {}}
  testID="ContainerView"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#e1e8ee",
          "borderRadius": 8,
          "flexDirection": "row",
        },
        Object {
          "borderColor": "#C3D1DD",
        },
      ]
    }
    testID="InputView"
  >
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholderTextColor="#86939e"
      style={
        Object {
          "color": "#000000",
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "fontSize": 16,
          "overflow": "hidden",
          "padding": 10,
        }
      }
      testID="Input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
